### PR TITLE
Ross api new further

### PR DESCRIPTION
**Pull request by rosswhitaker**
This new version includes the logging capability, fixes some instabilities/bugs, and removes all of the print statements.  There are some dangling and inappropriate comments that needed to be edited and sorted out.

### CODE REVIEW

File: aqandu/__init__.py [hunk 1]
```diff
@@ -1,26 +1,35 @@
...
+
 
 app = Flask(__name__)
 app.config.from_object(config)
-app.config["CACHE_TYPE"] = "simple"
-app.config["CACHE_DEFAULT_TIMEOUT"] = 1
-init(app)
-
-cache = Cache(app)
+assets.init(app)
```
Comment (JackWilb) [inline aqandu/__init__.py:25]: Please keep the caching
```suggestion
app.config["CACHE_TYPE"] = "simple"
app.config["CACHE_DEFAULT_TIMEOUT"] = 1
cache = Cache(app)

assets.init(app)
```

File: aqandu/gaussian_model_utils.py [hunk 2]
```diff
@@ -41,46 +52,58 @@ def convertToTimeCoordinatesVector(dates, time_offset):
...
 
     time_coordinates = [bin_number - lowest_bin_number for bin_number in time_coordinates]
-#    print("time1")
-#    print(time_coordinates)
     time_coordinates.sort()
-#    print("time2")
+    
+#    print("time1")
 #    print(time_coordinates)
+#    time_coordinates.sort()
```
Comment (JackWilb) [inline aqandu/gaussian_model_utils.py:85]: This can be removed or updated to for logging
```suggestion
```

File: aqandu/gaussian_model_utils.py [hunk 3]
```diff
@@ -149,36 +174,57 @@ def saveMatrixToFile(matrix, filename):
...
     for row in matrix:
-        row_index += 1
         prevValueIndex = None
+        this_fail = False
         for i in range(row.shape[0]):
             if row[i] != bad_value:
                 if prevValueIndex is None:
                     prevValueIndex = i
+# this takes care of the boundary at the beginning of the time sequence
+                    if (i > 0) and (i < SENSOR_INTERPOLATE_DISTANCE):
```
Comment (JackWilb) [inline aqandu/gaussian_model_utils.py:186]: Looks like the indentation is off a little here
```suggestion
                    # this takes care of the boundary at the beginning of the time sequence
                    if (i > 0) and (i < SENSOR_INTERPOLATE_DISTANCE):
```

File: aqandu/gaussian_model_utils.py [hunk 4]
```diff
@@ -234,8 +281,11 @@ def setupDataMatrix(sensor_data, space_coordinates, time_coordinates, device_loc
 #    saveMatrixToFile(data_matrix, '2interpolated.txt')
     data_matrix, space_coordinates = removeBadSensors(data_matrix, space_coordinates, 0.6)
 #    saveMatrixToFile(data_matrix, '3matrix_removed_bad.txt')
-#    data_matrix, time_coordinates = trimEdgeZeroElements(data_matrix, time_coordinates)
-#    saveMatrixToFile(data_matrix, '4matrixtrimmed.txt')
+
+#  This is important because bins on either end of the time range might not have enough measurements
+#    data_matrix, time_coordinates = trimBadEdgeElements(data_matrix, time_coordinates)
+
+    #    saveMatrixToFile(data_matrix, '4matrixtrimmed.txt')
```
Comment (JackWilb) [inline aqandu/gaussian_model_utils.py:288]: Looks like there is a ton of commented out code in this section, can we remove it?

File: aqandu/gaussian_model_utils.py [hunk 5]
```diff
@@ -313,12 +369,14 @@ def setupDataMatrix2(sensor_data, space_coordinates, time_coordinates, device_lo
...
     data_mask = (data_matrix != bad_value)
     data_counts = numpy.sum(data_mask, 0)
-    if (float(numpy.min(data_counts))/float(data_matrix.shape[0]) < 0.75):
-        print("WARNING: got time slice with too few data sensor values with value " + str(float(numpy.min(data_counts))/float(data_matrix.shape[0])) + " and index "  + str(numpy.nonzero((data_counts/data_matrix.shape[0]) < 0.75)))
-    time_averages = numpy.divide(numpy.sum(numpy.multiply(data_matrix,data_mask), 0), data_counts)
+    if (float(numpy.min(data_counts))/float(data_matrix.shape[0]) < TIME_SLICE_MIN_SENSOR_RATE):
+        logging.warn("WARNING: got time slice with too few data sensor values with value " + str(float(numpy.min(data_counts))/float(data_matrix.shape[0])) + " and index "  + str(numpy.nonzero((data_counts/data_matrix.shape[0]) < 0.75)))
+    sum_tmp = numpy.sum(numpy.multiply(data_matrix,data_mask), 0)
+    time_averages = numpy.divide(sum_tmp, data_counts, out=numpy.zeros_like(sum_tmp), where=(data_counts!=0))
+#    print("time_averages is " + str(time_averages))
```
Comment (JackWilb) [inline aqandu/gaussian_model_utils.py:379]: Remove or log
```suggestion
```

File: aqandu/api_routes.py [hunk 6]
```diff
@@ -227,9 +228,9 @@ def getEstimateMap():
 
     # step 2, load up length scales from file
     length_scales = utils.loadLengthScales('length_scales.csv')
-    print(f'Loaded {len(length_scales)} length scales.')
+#    print(f'Loaded {len(length_scales)} length scales.')
 
-    print('Loaded length scales:', length_scales, '\n')
+#    print('Loaded length scales:', length_scales, '\n')
```
Comment (JackWilb) [inline aqandu/api_routes.py:233]: Remove or log

File: aqandu/api_routes.py [hunk 7]
```diff
@@ -319,29 +319,13 @@ def getEstimateMap():
...
     locations_lon = locations_lon.flatten()
     elevations = elevations.flatten()
-    # print("C")
-    # print(locations_lat)
-    # print(locations_lon)
-
-    # print("D")
-    # print(locations_lat.reshape((lat_size, lon_size)))
-    # print(locations_lon.reshape((lat_size, lon_size)))
-
```
Comment (JackWilb) [inline aqandu/api_routes.py:344]: Nice!

File: aqandu/api_routes.py [hunk 8]
```diff
@@ -590,7 +572,7 @@ def request_model_data():
 
     start_date = query_parameters.get('start_date')
     end_date = query_parameters.get('end_date')
-    print("model requuest api with " + str(lat) + ":" + str(lon) + " and radius " + str(radius) + " and start " + str(start_date) + " and end " + str(end_date))
+#    print("model requuest api with " + str(lat) + ":" + str(lon) + " and radius " + str(radius) + " and start " + str(start_date) + " and end " + str(end_date))
```
Comment (JackWilb) [inline aqandu/api_routes.py:575]: Remove or log

File: aqandu/api_routes.py [hunk 9]
```diff
@@ -805,28 +670,58 @@ def getEstimatesForLocations():
...
     query_end_datetime = utils.parseDateString(query_end_date)
 
-    print((
-        f"Query parameters: lat={query_lats} lon={query_lons} start_date={query_start_datetime}"
-        f" end_date={query_end_datetime} estimaterate={query_rate}"
-    ))
+#    print((
+#        f"Query parameters: lat={query_lats} lon={query_lons} start_date={query_start_datetime}"
+#        f" end_date={query_end_datetime} estimaterate={query_rate}"
+#    ))
```
Comment (JackWilb) [inline aqandu/api_routes.py:676]: Can this be removed?

File: aqandu/api_routes.py [hunk 10]
```diff
@@ -805,28 +670,58 @@ def getEstimatesForLocations():
...
-    print(f'Loaded {len(correction_factors)} correction factors.')
+#    print(f'Loaded {len(correction_factors)} correction factors.')
 
     # step 2, load up length scales from file
     length_scales = utils.loadLengthScales('length_scales.csv')
-    print(f'Loaded {len(length_scales)} length scales.')
+#    print(f'Loaded {len(length_scales)} length scales.')
 
-    print('Loaded length scales:', length_scales, '\n')
+#    print('Loaded length scales:', length_scales, '\n')
```
Comment (JackWilb) [inline aqandu/api_routes.py:724]: Log or remove

File: aqandu/api_routes.py [hunk 11]
```diff
@@ -839,9 +734,9 @@ def getEstimatesForLocations():
     elevation_length_scale = length_scales[0]['elevation']
     time_length_scale = length_scales[0]['time']
 
-    print(
-        f'Using length scales: latlon={latlon_length_scale} elevation={elevation_length_scale} time={time_length_scale}'
-    )
+#    print(
+#        f'Using length scales: latlon={latlon_length_scale} elevation={elevation_length_scale} time={time_length_scale}'
+#    )
```
Comment (JackWilb) [inline aqandu/api_routes.py:739]: Log or remove

File: aqandu/api_routes.py [hunk 12]
```diff
@@ -887,10 +781,10 @@ def getEstimatesForLocations():
     for datum in sensor_data:
         datum['type'] =  sensor_source_to_type[datum['SensorSource']]
 
-    print(f'Fields: {sensor_data[0].keys()}')
+    app.logger.info(f'Fields: {sensor_data[0].keys()}')
 
     # step 4.5, Data Screening
-    print('Screening data')
+#    print('Screening data')
```
Comment (JackWilb) [inline aqandu/api_routes.py:787]: Log or remove

File: aqandu/api_routes.py [hunk 13]
```diff
@@ -903,28 +797,41 @@ def getEstimatesForLocations():
...
-        sensor_data, latlon_length_scale, elevation_length_scale, time_length_scale)
-    # step 9, Compute estimates
-    yPred, yVar = gaussian_model_utils.estimateUsingModel(
-        model, query_lats, query_lons, query_elevations, query_dates, time_offset)
+    # # step 8, Create Model
+    # model, time_offset = gaussian_model_utils.createModel(
+    #     sensor_data, latlon_length_scale, elevation_length_scale, time_length_scale)
+    # # step 9, Compute estimates
+    # yPred, yVar = gaussian_model_utils.estimateUsingModel(
+    #     model, query_lats, query_lons, query_elevations, query_dates, time_offset)
```
Comment (JackWilb) [inline aqandu/api_routes.py:807]: Can this be removed?

File: aqandu/api_routes.py [hunk 14]
```diff
@@ -903,28 +797,41 @@ def getEstimatesForLocations():
...
-    for i in range(num_times):
-        estimates.append(
-            {'PM2_5': (yPred[:,i]).tolist(), 'variance': (yVar[:,i]).tolist(), 'datetime': query_dates[i].strftime('%Y-%m-%d %H:%M:%S%z'), 'Latitude': query_lats.tolist(), 'Longitude': query_lons.tolist(), 'Elevation': query_elevations.tolist()}
-            )
+    yPred = np.empty((num_locations, 0))
+    yVar = np.empty((num_locations, 0))
+    for i in range(len(query_sequence)):
+    # step 7, Create Model
+#        print(sensor_sequence[i])
+#        print(query_sequence[i])
```
Comment (JackWilb) [inline aqandu/api_routes.py:819]: Log or remove

**Review by JackWilb — CHANGES_REQUESTED** [review]
Thanks for the smaller number of line changes, it made the review much easier. I'm assuming that you've tested the code locally, and I've just made some comments on lines that seem to have been missed for removal or logging. Once you clean these up, I'll be happy to approve 😄

File: aqandu/gaussian_model_utils.py (hunk 4 above)
Comment (rosswhitaker) [reply aqandu/gaussian_model_utils.py:285]: Agreed, it should be removed.  However, I would like to come back to that as I continue to test and understand it better.  It sits in the flow right now to remind me (for the time being) of how this part used to go.

File: aqandu/api_routes.py (hunk 11 above)
Comment (rosswhitaker) [reply aqandu/api_routes.py:739]: Done.

File: aqandu/api_routes.py (hunk 10 above)
Comment (rosswhitaker) [reply aqandu/api_routes.py:724]: Done.

File: aqandu/api_routes.py (hunk 13 above)
Comment (rosswhitaker) [reply aqandu/api_routes.py:807]: This is another case where the code is a work in progress, and until I have more testing done, I want to be able to refer to this.  As this settles we can remove it.  I put a comment above explaining that.

File: aqandu/api_routes.py (hunk 14 above)
Comment (rosswhitaker) [reply aqandu/api_routes.py:819]: Done.

File: aqandu/gaussian_model_utils.py (hunk 4 above)
Comment (rosswhitaker) [reply aqandu/gaussian_model_utils.py:285]: I will put a note explaining to remove this after fully tested.

Comment (rosswhitaker) [reply aqandu/gaussian_model_utils.py:285]: Much of the commented code is now removed.  More to come.

**Review by JackWilb — APPROVED** [review]
Looks good generally. There are some unaddressed comments from my previous review, but we're most of the way there. 

Another note, caching seems to have been removed somehow, so I'll need to go back in and fix that up before we do a proper deployment. With these changes are you ready for a deployment?

**Comment by rosswhitaker** [comment]
This fixes a few remaining issues and addresses some numerical problems.  FlaskCache should now be working.

**Review by JackWilb — APPROVED** [review]

**Comment by rosswhitaker** [comment]
Thanks for all the help/patience, Jack.  Let me know if you want to work together on updating that contours code to use the new numerics/api.   Shouldn't be too hard – although some parts of that code I don't full understand.

**Pull request merged**